src/zcash/History.cpp

```diff
@@ -0,0 +1,122 @@
+#include "zcash/History.hpp"
+
+#include <stdexcept>
+
+#include <boost/foreach.hpp>
+
+#include "serialize.h"
+#include "streams.h"
+#include "uint256.h"
+#include "librustzcash.h"
+
+namespace libzcash {
+
+void HistoryCache::Extend(const HistoryNode &leaf) {
+    appends[length++] = leaf;
+}
+
+void HistoryCache::Truncate(HistoryIndex newLength) {
+    for (HistoryIndex idx = length; idx > newLength; idx--) {
+        appends.erase(idx);
+    }
+
+    length = newLength;
+    // we track how deep updates go back in the tree, so we could later
+    // update everything starting from `updateDepth`
+    //
+    // imagine we rolled two blocks back and then put another 3 blocks on top
+    // of the rolled back state. In that case `updateDepth` will be H-3, while length
+    // will be H (where H is a final chain height after such operation). So we know that
+    // history entries in the range of H-3..H are expected to be pushed into the database
+    // to replace/append to the persistent nodes there.
+    if (updateDepth > length) updateDepth = length;
+}
+
+HistoryNode NewNode(
+        uint256 subtreeCommitment,
+        uint32_t startTime,
+        uint32_t endTime,
+        uint32_t startTarget,
+        uint32_t endTarget,
+        uint256 startSaplingRoot,
+        uint256 endSaplingRoot,
+        uint256 subtreeTotalWork,
+        uint64_t startHeight,
+        uint64_t endHeight,
+        uint64_t saplingTxCount
+    )
+{
+    CDataStream buf(SER_DISK, 0);
+    HistoryNode result;
+
+    buf << subtreeCommitment;
+    buf << startTime;
+    buf << endTime;
+    buf << startTarget;
+    buf << endTarget;
+    buf << startSaplingRoot;
+    buf << endSaplingRoot;
+    buf << subtreeTotalWork;
+    buf << COMPACTSIZE(startHeight);
```

Comment:
I am probably misunderstanding something, but `serialize_compact_uint` isn't defined in the ZIP, but it is a dynamic encoding that has a `MAX_SIZE = 33554432` (which is _maybe_ not a problem) but also varies in length. Won't this cause issues if the heights or `saplingTxCount` changes dramatically? (For heights, maybe regtest, for `saplingTxCount` I guess a block with many Sapling transactions)

Reply:
Nevermind! I see the part of the ZIP which discusses this as well as the thread with daira about using fixed-length encodings.